### DISCUSSION TITLE
[Technical debt] Úložiště avatarů

### DISCUSSION THREAD

**Discussion by srdecny** [cityvizor/cityvizor#349] [General]
Aktuální návrh aplikace je založen na tom, že `server` a `worker` mezi sebou sdílí data nutná pro import. Toto se aktálně děje přes [bind-mounting](https://github.com/cityvizor/cityvizor/blob/master/docker-compose.yml#L27) sdílené složky `/data` mezi oběma kontejnery přes docker-compose. [Do této složky se ale ukládají i avatary](https://github.com/cityvizor/cityvizor/blob/master/server/src/config/index.ts#L39) (erby) jednotlivých obcí. Jedná se tedy o další persistentní data, která je nutno zálohovat. Co s tím?

- Buď to necháme tak, jak je teď a kromě DB budem zálohovat ještě i tento adresář. 
- Nebo si ušetříme práci a avatary dáme do DB (nevím, jak moc je to dobrý nápad) a sdílený adresář zálohovat nebudeme. V tom případě se vystavujeme riziku, že přijdeme o data, která čekají ve frontě na import (uložená serverem, ale worker je ještě nezpracoval). Tohle kritické okno je celkem malé a v řádu sekund, za běžných podmínek (pokud se import worker třeba nezasekne).
- Nebo se avatary budou sypat na nějaké jiné statické úložiště (třeba pod CDNkou, kdysi se experimentovalo s myšlenkou, že by se dávaly do AWS S3 bucketu)
-  Rozšíříme tenhle dataset: https://github.com/cesko-digital/obce o erby. Ale nevím, jestli se nám podaří vyscrapovat fakt všechny obce (a asi spíš ne). A asi je to blbost, protože příspěvkové organizace budou mít svá vlastní loga.

Thoughts?

**Comment by smarek** [comment]
1. Erby / Avatary jsou jednoduché a malé soubory, které by klidně v databázi být mohly, ale není to imho koncepčně dobrý nápad
2. Scrapování veřejných dat nikdy nenaplní potřeby uživatelů s názorem
3. Je potřeba vyřešit separaci souborů od modulů, šlo by sice udělat z worker+(admin část client) samostatně funkční celek, který by si držel interně i soubory, ale nebylo by to škálovatelné

Takže rozhodnutí je použít nějaký S3-compatible storage, nejsem příznivec AWS S3 ani DigitalOcean Spaces, ale obě řešení jsou S3 kompatibilní a lze je použít jak pro interní (private) object/file storage, tak jako CDN.

Interně chceme open-source, ideálně ve formě docker modulu do stávající architektury (stejně jako máme nyní postgres)

Na první google nástřel jsem našel
  - https://min.io/
  - https://www.exoscale.com/object-storage/
  - https://github.com/ceph/cn
  - příp. rozcestník https://rclone.org/s3/

@sorki 
1. prosím rozšiř aktuální architekturu o storage
  1. hledáme open-source se správnou licencí, která je kompatibilní se způsobem provozu našich aplikací v OM
  2. Cílem je aby se součástí ENV konfigurace "server" a "worker" modulů byl odkaz na s3 storage a nastavení (id + secret), tak aby se nechalo provozovat Docker-only nebo pokud daná organizace chce, mohla použít AWS nebo DigitalOcean nebo jiný S3-kompatibilní úložiště.
  3. Případně pokud ten vybraný storage bude dostupný i jako CDN, tak s konfigurací front-end doménového jména a schopností ty "public" soubory sypat ven přímo uživateli, aby se aplikace mohly s jednotlivými public soubory odkazovat přímo na CDN

**Comment by smarek** [comment]
Tak @sorki zařídil min.io na adrese minio.cityvizor.cz takže ti @srdecny můžu udělat přístup a můžeme operace nad FS převést do operací nad S3 storage, jen asi při cron/worker/import-queue by to chtělo vytvářet temp adresář třeba `{id importu}-{unix timestamp}` aby si paralelně spuštěné procesy nelezly do zelí v rámci FS
